snap/hooks/configure

```diff
@@ -1,158 +1,84 @@
-        mv "${kiosk_display}"  "${kiosk_display}.save"
-        let display_changes+=1
-      fi
-    fi
-  fi
-fi
-
-# display-layout
-display_layout=$(snapctl get display-layout)
-if [ -n "${display_layout}" ]; then
-  if [ -e "${kiosk_display}" ]; then
-    if [ -z "$(grep "^  ${display_layout}:" "${kiosk_display}")" ]; then
-      echo "ERROR: '$display_layout' is not a layout in ${kiosk_display}:"
-      echo "---"
-      cat "${kiosk_display}"
-      echo "---"
-      exit 1
-    fi
-  else
-    if [ "${display_layout}" != "default" ]; then
-      echo "ERROR: '$display_layout' is not a layout in ${kiosk_display}:"
-      echo "---"
-      cat "${kiosk_display}"
-      echo "---"
-      exit 1
-    fi
-  fi
-
-  if [ "display-layout=${display_layout}" != "$(grep \^display-layout= ${kiosk_config})" ]; then
-    if [ $config_changes -gt 0 ]; then
-      sed '/^display-layout=/d' ${kiosk_config}
-    else
-      sed --in-place=.save '/^display-layout=/d' ${kiosk_config}
-    fi
-    echo display-layout=${display_layout} >> ${kiosk_config}
-    let config_changes+=1
-  fi
-fi
+# config_file
+# A kiosk configuration can be supplied through the kiosk-config option
+kiosk_config="$(snapctl get kiosk-config)"
+config_temp="$(mktemp)"
+cat <<EOT > "${config_temp}"
+# DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN
+# USE 'snap set $SNAP_INSTANCE_NAME kiosk-config=...' INSTEAD
 
-# vt
-vt=$(snapctl get vt)
-if [ -n "${vt}" ]; then
-  if [ -n "${vt//[0-9]}" ] || [ ! -e "/dev/tty$vt" ]; then
-    echo "ERROR: '$vt' is not a valid VT"
-    exit 1
-  fi
+# ** Options supplied through kiosk-config (begin) **
+${kiosk_config:-# (none)}
+# ** Options supplied through kiosk-config (end) **
+EOT
 
-  if [ "vt=${vt}" != "$(grep vt= ${kiosk_config})" ]; then
-    if [ $config_changes -gt 0 ]; then
-      sed '/^vt/d' ${kiosk_config}
-    else
-      sed --in-place=.save '/^vt/d' ${kiosk_config}
-    fi
-    echo vt=${vt} >> ${kiosk_config}
-    let config_changes+=1
-  fi
-fi
+config_entry() {
+  if ! grep -q "^$1=" "${config_temp}"; then
+    cat <<EOT >> "${config_temp}"
 
-# cursor
-cursor=$(snapctl get cursor)
-if [ -n "${cursor}" ]; then
-  if [[ ! "${cursor}" =~ ^(auto|none|software)$ ]]; then
-    echo "ERROR: '$cursor' is not a valid cursor option (auto|none|software)"
-    exit 1
+# $3
+$1=$2
+EOT
   fi
+}
 
-  if [ "${cursor}" == "none" ]; then
-    cursor="null"
-  fi
+if grep -q snap_core= /proc/cmdline && grep -q vt.handoff= /proc/cmdline
+then
+  cat <<EOT >> "${config_temp}"
 
-  if [ "cursor=${cursor}" != "$(grep cursor= ${kiosk_config})" ]; then
-    if [ $config_changes -gt 0 ]; then
-      sed '/^cursor/d' ${kiosk_config}
-    else
-      sed --in-place=.save '/^cursor/d' ${kiosk_config}
-    fi
-    echo cursor=${cursor} >> ${kiosk_config}
-    let config_changes+=1
-  fi
+# WORKAROUND: If 'vt.handoff' is set, Mir can only start with vt=0.
+# The correct fix is: https://github.com/snapcore/pi-gadget/pull/12
+vt=0
+EOT
 fi
```

Comment:
Is `vt.handoff` alone that triggers this? Should you check for `snap_core` then?

Reply:
That's the logic in mir-kiosk, I'll dig to find out how we got there.